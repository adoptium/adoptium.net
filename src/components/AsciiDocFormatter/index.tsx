'use client';

import React from 'react';
import { Link } from "@/i18n/navigation";
import { useTranslations } from 'next-intl';
import parse, { domToReact, HTMLReactParserOptions, Element, DOMNode } from "html-react-parser";
import AccordionItem from "@/components/Asciidoc/AccordionItem";

interface AsciiDocFormatterProps {
  content: string;
}

const AsciiDocFormatter: React.FC<AsciiDocFormatterProps> = ({ content }) => {
  const isBrowser = typeof window !== "undefined";
  const t = useTranslations("Asciidoc");

  const options: HTMLReactParserOptions = {
    replace: (node) => {
      if (!(node instanceof Element)) {
        return undefined;
      }

      // Transform <a> tags
      if (node.name === "a" && node.attribs) {
        let isExternal = false;
        if (isBrowser && node.attribs.href) {
          isExternal =
            node.attribs.href.startsWith("http") &&
            !node.attribs.href.includes(window.location.host);
        }
        if (isExternal) {
          return (
            <a
              href={node.attribs.href}
              target="_blank"
              rel="noopener noreferrer"
              className={node.attribs.class}
            >
              {domToReact(node.children as DOMNode[], options)}
              <i className="fa fa-external-link fa-xs p-1" />
            </a>
          );
        } else {
          return (
            <Link href={node.attribs.href} className={node.attribs.class}>
              {domToReact(node.children as DOMNode[], options)}
            </Link>
          );
        }
      }

      // Transform <i> tags
      if (
        node.name === "i" &&
        node.attribs?.class &&
        node.attribs.class.includes("fa-docker")
      ) {
        return <i className={node.attribs.class.replace("fa", "fab")} />;
      }

      // Transform <td> tags
      if (
        node.name === "td" &&
        node.attribs?.class === "icon"
      ) {
        return (
          <td>
            {domToReact(node.children as DOMNode[], options)}
            <i className="fa fa-circle-info fa-xl" aria-hidden="true" />
          </td>
        );
      }

      // Transform Table of Contents
      if (
        node.name === "div" &&
        node.attribs?.class === "toc"
      ) {
        // Get the ul element class="sectlevel1"
        const tocList = node.children.find(
          (child) =>
            child instanceof Element &&
            child.attribs &&
            child.attribs.class &&
            typeof child.attribs.class === 'string' &&
            child.attribs.class.includes("sectlevel1")
        ) as Element | undefined;

        return (
          <div className="my-6 rounded-lg shadow-md overflow-hidden border border-gray-300 dark:border-gray-700">
            <details className="group">
              <summary className="flex items-center justify-between p-4 bg-gradient-to-r from-gray-100 to-gray-200 dark:from-gray-800 dark:to-gray-700 cursor-pointer hover:from-gray-200 hover:to-gray-300 dark:hover:from-gray-700 dark:hover:to-gray-600 transition-all duration-300">
                <div className="flex items-center">
                  <i className="fa fa-list-ul mr-3 text-pink dark:text-purple-400" aria-hidden="true" />
                  <span className="font-semibold text-lg text-gray-800 dark:text-gray-200">
                    {t('table-of-contents')}
                  </span>
                </div>
                <i className="fa fa-chevron-down text-pink dark:text-purple-400 group-open:rotate-180 transition-transform duration-300" aria-hidden="true" />
              </summary>
              <div className="p-4 bg-gray-50 dark:bg-gray-800">
                {tocList ? (
                  <div className="toc-container pl-2 not-prose [&_a]:text-gray-700 [&_a:hover]:text-gray-900 dark:[&_a]:text-gray-200 dark:[&_a:hover]:text-white">
                    <ul className="space-y-1">
                      {domToReact(tocList.children as DOMNode[], options)}
                    </ul>
                  </div>
                ) : null}
              </div>
            </details>
          </div>
        );
      }

      // Transform iframe tags
      if (node.name === "iframe") {
        return (
          <iframe {...node.attribs} className="w-full aspect-video" height="400">
            {domToReact(node.children as DOMNode[], options)}
          </iframe>
        );
      }

      // Transform <details> and <summary> tags into Accordion
      if (node.name === "details") {
        const summary = node.children.find(
          (child) =>
            child instanceof Element &&
            child.name === "summary"
        ) as Element | undefined;

        const summaryContent = summary
          ? domToReact(summary.children as DOMNode[], options)
          : "Details";

        const detailsContent = node.children.filter(
          child => !(child instanceof Element && child.name === "summary")
        );

        return (
          <AccordionItem title={summaryContent}>
            {domToReact(detailsContent as DOMNode[], options)}
          </AccordionItem>
        );
      }

      // Default case - return undefined to keep original element
      return undefined;
    }
  };

  return (
    <div className="prose prose-invert lg:prose-lg max-w-none">
      {parse(content, options)}
    </div>
  );
};

export default AsciiDocFormatter;
