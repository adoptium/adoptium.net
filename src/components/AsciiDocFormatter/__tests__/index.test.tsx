import React from "react"
import { render, screen } from "@testing-library/react"
import { describe, expect, it, vi } from "vitest"
import AsciiDocFormatter from ".."

// Mock next-intl's useTranslations hook
vi.mock("next-intl", () => ({
  useTranslations: () => (key: string) => {
    const translations: Record<string, string> = {
      "table-of-contents": "Table of Contents"
    };
    return translations[key] || key;
  }
}));

// Mock AccordionItem component
vi.mock("@/components/Asciidoc/AccordionItem", () => ({
  __esModule: true,
  default: ({ title, children }: { title: React.ReactNode; children: React.ReactNode }) => (
    <div data-testid="mock-accordion">
      <div data-testid="accordion-title">{title}</div>
      <div data-testid="accordion-content">{children}</div>
    </div>
  ),
}))

describe("AsciiDocFormatter", () => {
  it("renders basic HTML content correctly", () => {
    const { container } = render(<AsciiDocFormatter content="<p>Hello world</p>" />)
    expect(screen.getByText("Hello world")).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it("transforms internal links correctly", () => {
    const content = '<p><a href="/internal-link">Internal Link</a></p>'
    render(<AsciiDocFormatter content={content} />)
    
    const link = screen.getByText("Internal Link")
    expect(link.tagName).toBe("A")
    expect(link).toHaveAttribute("href", "/internal-link")
    expect(link).not.toHaveAttribute("target")
    expect(link).not.toHaveAttribute("rel")
  })

  it("transforms external links correctly with external icon", () => {
    // Mock window.location.host
    Object.defineProperty(window, "location", {
      value: { host: "adoptium.net" },
      writable: true
    })

    const content = '<p><a href="https://example.com">External Link</a></p>'
    const { container } = render(<AsciiDocFormatter content={content} />)
    
    const link = screen.getByText("External Link")
    expect(link.tagName).toBe("A")
    expect(link).toHaveAttribute("href", "https://example.com")
    expect(link).toHaveAttribute("target", "_blank")
    expect(link).toHaveAttribute("rel", "noopener noreferrer")
    
    // Check for the external link icon
    const icon = container.querySelector(".fa-external-link")
    expect(icon).toBeInTheDocument()
  })

  it("transforms FontAwesome Docker icon correctly", () => {
    const content = '<p><i class="fa fa-docker"></i></p>'
    const { container } = render(<AsciiDocFormatter content={content} />)
    
    const icon = container.querySelector(".fab.fa-docker")
    expect(icon).toBeInTheDocument()
  })

  it("transforms table of contents correctly", () => {
    const content = `
      <div class="toc">
        <ul class="sectlevel1">
          <li><a href="#section1">Section 1</a></li>
          <li><a href="#section2">Section 2</a></li>
        </ul>
      </div>
    `
    render(<AsciiDocFormatter content={content} />)
    
    // Check for Table of Contents text
    expect(screen.getByText("Table of Contents")).toBeInTheDocument()
    
    // Check for section links
    expect(screen.getByText("Section 1")).toBeInTheDocument()
    expect(screen.getByText("Section 2")).toBeInTheDocument()
  })

  it("transforms iframe tags correctly", () => {
    const content = '<iframe src="https://example.com/embed" title="Sample Embed"></iframe>'
    const { container } = render(<AsciiDocFormatter content={content} />)
    
    const iframe = container.querySelector("iframe")
    expect(iframe).toBeInTheDocument()
    expect(iframe).toHaveAttribute("src", "https://example.com/embed")
    expect(iframe).toHaveAttribute("title", "Sample Embed")
    expect(iframe).toHaveClass("w-full")
    expect(iframe).toHaveClass("aspect-video")
  })

  it("transforms details/summary into AccordionItem", () => {
    const content = `
      <details>
        <summary>Accordion Title</summary>
        <p>Accordion Content</p>
      </details>
    `
    render(<AsciiDocFormatter content={content} />)
    
    // Check that the AccordionItem mock is used
    expect(screen.getByTestId("mock-accordion")).toBeInTheDocument()
    
    // Check that the title and content are passed correctly
    const title = screen.getByTestId("accordion-title")
    expect(title).toHaveTextContent("Accordion Title")
    
    const content_ = screen.getByTestId("accordion-content")
    expect(content_).toHaveTextContent("Accordion Content")
  })

  it("transforms icon td cells correctly", () => {
    const content = '<table><tbody><tr><td class="icon"></td></tr></tbody></table>'
    const { container } = render(<AsciiDocFormatter content={content} />)
    
    const icon = container.querySelector(".fa-circle-info")
    expect(icon).toBeInTheDocument()
  })
})
