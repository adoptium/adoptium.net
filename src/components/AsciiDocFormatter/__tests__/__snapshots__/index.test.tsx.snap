// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AsciiDocFormatter > renders basic HTML content correctly 1`] = `
<div>
  <div
    class="prose prose-invert lg:prose-lg max-w-none"
  >
    <p>
      Hello world
    </p>
  </div>
</div>
`;
